compiler/rustc_codegen_ssa/src/mir/place.rs

```diff
@@ -387,13 +388,21 @@ impl<'a, 'tcx, V: CodegenObject> PlaceRef<'tcx, V> {
                     let niche_llty = bx.cx().immediate_backend_type(niche.layout);
                     let niche_value = variant_index.as_u32() - niche_variants.start().as_u32();
                     let niche_value = (niche_value as u128).wrapping_add(niche_start);
-                    // FIXME(eddyb): check the actual primitive type here.
-                    let niche_llval = if niche_value == 0 {
-                        // HACK(eddyb): using `c_null` as it works on all types.
-                        bx.cx().const_null(niche_llty)
-                    } else {
-                        bx.cx().const_uint_big(niche_llty, niche_value)
+
+                    let BackendRepr::Scalar(scalar) = niche.layout.backend_repr else {
+                        bug!("expected a scalar placeref for the niche");
                     };
+                    let max = niche.layout.size.unsigned_int_max();
+                    // Niche ranges that wrap around may be represented as out-of-range values
+                    // e.g. niche at 0u8 may end up being u8::MAX + 1.
```

Comment:
I don't think that's true? AFAIK our niche ranges must always be inbounds of the scalar type they refer to. However, above you are doing `u128` arithmetic so even if the niche range endpoints are inbounds, your `niche_value` is not.

Reply:
niche_start is inbounds, but niche_start + niche_value not. And this was passed as such to the backend, i.e. I have seen 256 passed to llvm for a repr(i8) enum via `const_uint_big`.

Reply:
`niche_start + niche_value` ought to be computed wrapping on the type of the niche field. But the code here takes a shortcut, but doing `u128`  arithmetic instead. *That* is why it can go out-of-bounds.

Reply:
> niche_start + niche_value ought to be computed wrapping on the type of the niche field

If that were the case then it would have been buggy before. Note that I did not change the calculation and the out of range value was passed to the backend. This worked for integers. What I'm fixing is not that case but that llvm didn't like them for pointers.

Reply:
> And the result was passed as such to the backend, i.e. I have seen 256 passed to llvm for a repr(i8) enum via const_uint_big.

If that worked before, that means LLVM has been doing the truncation for integers. It just didn't like doing them for pointers, I guess.

Reply:
What your PR implements is exactly "do the arithmetic wrapping on the size of the niche field", though, so I don't understand what you are disagreeing with.

Reply:
Yes, that's also my assumption. Which is why my comment is trying to explain that we do this to cover the pointer case which isn't handled by the backend (but for other cases they do handle it).

Reply:
Your comment says that niche values can go out-of-bounds.  That's just not correct. The correct niche value is computed by wrapping arithmetic on the size of the niche field and that cannot go out-of-bounds.

It'd be more clear to make this `let niche_value = (niche_value as u128).wrapping_add(niche_start) & max;`, so that the wrapping arithmetic is done in a single proper step.

Reply:
What I don't understand is why you think my comment is incorrect.

We do calculate out of bound values. The backend accepts out of bound values in the common case. So we could just pass them through, as we did before. Ranges that wrap around resulting in out of bounds values is exactly what happens.

The only reason to do the wrapping ourselves is the special case that's currently only reachable via unstable attributes.

Reply:
This calculates the niche value we tell the backend to set. Not the value we use internally to represent the niche.

The value we tell the backend can go out of bounds due to truncation in the backend.
I could have special-cased this to just do the wrapping outselves for pointers. But feels silly. Since they're equivalent I do it for integers too. But that's not a necessary property of the code.

Reply:
Would this be better?

> The calculated niche_value may end up being out of bounds. This would be fine for integer types since such integer constants do get truncated by the backend to fit the type. But to also cover pointer types we always do it ourselves.

Reply:
I think your comment is misleading. The reason the value can be out-of-bounds is that it wasn't computed the right way. Or rather, it wasn't yet fully computed. The comment occurs half-way through the computation.

Reply:
I don't see why we would talk about the backend truncation behavior... it's a bit unclean to rely on that anyway (reasonably a backend could complain if we build an `i8` literal of value 256), and we are not relying on it any more with this PR.